### ISSUE TITLE
Fail to use tf.transpose() after tf.nn.embedding_lookup() and tf.layers.conv1d() while build a tflite file

### ISSUE THREAD

**Issue by lost-libra** [tensorflow/tensorflow#27781]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04.3
- TensorFlow installed from (source or binary): source
- TensorFlow version (or github SHA if from source): 1.13.1

**Provide the text output from tflite_convert**
2019-04-12 16:14:22.935998: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 13 operators, 22 arrays (0 quantized)
2019-04-12 16:14:22.936163: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 13 operators, 22 arrays (0 quantized)
2019-04-12 16:14:22.939819: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 6 operators, 14 arrays (0 quantized)
2019-04-12 16:14:22.939890: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 2: 5 operators, 12 arrays (0 quantized)
2019-04-12 16:14:22.939945: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 3: 5 operators, 12 arrays (0 quantized)
2019-04-12 16:14:22.939972: F tensorflow/lite/toco/graph_transformations/propagate_fixed_sizes.cc:1605] Check failed: axis < input_shape.dimensions_count() (76044600 vs. 4)
Aborted (core dumped)
```
# Copy and paste here
```

Also, please include a link to a GraphDef or the model if possible.

**Any other info / logs**

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
`ipts = tf.placeholder(dtype=tf.int32, shape=(None, 50), name='x')`
`embedding = tf.get_variable(name='embedding', shape=[3000, 256], initializer=initializer)`
`res = tf.nn.embedding_lookup(params=embedding, ids=ipts)`
`res = tf.layers.conv1d(ipts, 100, 1)`
`res = tf.transpose(res, perm=[0, 2, 1])    # Here cause the BUG`
`sess = tf.Session()`
`sess.run(tf.global_variables_initializer())`
`converter = tf.contrib.lite.TFLiteConverter.from_session(sess, [ipts], [res])`
`tflite_model = converter.convert()`
`open('converted_model.tflite', 'wb').write(tflite_model)`

**Comment by lost-libra** [comment]
When I delete tf.transpose() function, the code runs well. Sometimes it also throws out this information: F tensorflow/lite/toco/graph_transformations/convert_trivial_transpose_to_reshape.cc:44] Check failed: new_major_index_ordering.size() == old_major_index_ordering.size() (2 vs. 3)
I really confused. Do anyone met this error before? Thanks.

**Comment by lost-libra** [comment]
When I delete tf.layers.conv1d() function or tf.nn.embedding_lookup() function, the code also runs well. When I delete one of tf.layers.conv1d(), tf.nn.embedding_lookup() and tf.transpose() function, it also runs well. Or even when I change the function order that put tf.transpose() in front of tf.layers.conv1d(), it also runs well. So it seems that I can't use codes as follow: tf.nn.embedding_lookup()->tf.layers.conv1d()->tf.transpose(). By the way, when I use tf.layers.dense() instead of tf.layers.conv1d(), the results are the same. Anyone can help? Thank you so much!

**Comment by haozha111** [comment]
I could convert the model with the following code:

def main(args):
  ipts = tf.placeholder(dtype=tf.float32, shape=(None, 50, 50), name='x')
  ids = tf.placeholder(dtype=tf.int32, shape=(None, 50), name='x')
  embedding = tf.get_variable(name='embedding', shape=[3000, 256], initializer=tf.initializers.random_uniform)
  res = tf.nn.embedding_lookup(params=embedding, ids=ids)
  res = tf.layers.conv1d(ipts, 100, 1)
  res = tf.transpose(res, perm=[0, 2, 1]) # Here cause the BUG
  sess = tf.Session()
  sess.run(tf.global_variables_initializer())
  converter = tf.lite.TFLiteConverter.from_session(sess, [ipts], [res])
  tflite_model = converter.convert()
  open('/tmp/converted_model.tflite', 'wb').write(tflite_model)

if __name__ == '__main__':
  app.run(main)

Probably sync to latest tf-nightly and try?

**Comment by lost-libra** [comment]
> I could convert the model with the following code:
> 
> def main(args):
> ipts = tf.placeholder(dtype=tf.float32, shape=(None, 50, 50), name='x')
> ids = tf.placeholder(dtype=tf.int32, shape=(None, 50), name='x')
> embedding = tf.get_variable(name='embedding', shape=[3000, 256], initializer=tf.initializers.random_uniform)
> res = tf.nn.embedding_lookup(params=embedding, ids=ids)
> res = tf.layers.conv1d(ipts, 100, 1)
> res = tf.transpose(res, perm=[0, 2, 1]) # Here cause the BUG
> sess = tf.Session()
> sess.run(tf.global_variables_initializer())
> converter = tf.lite.TFLiteConverter.from_session(sess, [ipts], [res])
> tflite_model = converter.convert()
> open('/tmp/converted_model.tflite', 'wb').write(tflite_model)
> 
> if **name** == '**main**':
> app.run(main)
> 
> Probably sync to latest tf-nightly and try?

Thank you for your answer and sorry for replying lately, but when I try your code after update tensorflow to 1.14.0.dev2019042301, it also doesn't work. Do I miss some important information? Or can you please show me your environment? By the way, I update tf-nightly by pip, and the BUG shows as: 2019-04-25 15:07:37.018200: F tensorflow/lite/toco/graph_transformations/propagate_fixed_sizes.cc:1813] Check failed: axis < input_shape.dimensions_count() (84489656 vs. 4)

**Comment by haozha111** [comment]
I can reproduce it under tf-nightly now. But with a simpler code:

def func():
  ipts = tf.placeholder(dtype=tf.float32, shape=(None, 50, 50), name='x')
  y = tf.layers.conv1d(ipts, filters=2, kernel_size=1)
  res = tf.transpose(y, perm=[0,2,1]) # Here cause the BUG
  sess = tf.Session()
  sess.run(tf.global_variables_initializer())

  converter = tf.lite.TFLiteConverter.from_session(sess, [ipts], [res])
  tflite_model = converter.convert()

func()

Also found some interesting things to bypass this error:
1) Change the 1st dim of ipts to something > 1.  or
2) use only one filter in conv1d. or
3) use another perm=[1,0,2]

Either of these could eliminate the error. Please stay tuned a bit as I continue to investigate the issue. Thanks.

**Comment by mtu** [comment]
I met the same problem @haozha111 . 
Hope you can fix this bug, thank you.

**Comment by terrykong** [comment]
Any update here? I've also hit this same issue

**Comment by lost-libra** [comment]
> Any update here? I've also hit this same issue

I tried tf nightly version 20190611, still doesn't work.

**Comment by paanguin** [comment]
I have 16 transpose operations in my tensorflow graph. I hope to figure out which one causes the problem, but tflite_convert doesn't provide that information.

**Comment by paanguin** [comment]
Is there any connection between this issue and following one?
https://github.com/tensorflow/tensorflow/issues/27640

**Comment by paanguin** [comment]
> I can reproduce it under tf-nightly now. But with a simpler code:
> 
> ```
> def func():
>   ipts = tf.placeholder(dtype=tf.float32, shape=(None, 50, 50), name='x')
>   y = tf.layers.conv1d(ipts, filters=2, kernel_size=1)
>   res = tf.transpose(y, perm=[0,2,1]) # Here cause the BUG
>   sess = tf.Session()
>   sess.run(tf.global_variables_initializer())
> 
>   converter = tf.lite.TFLiteConverter.from_session(sess, [ipts], [res])
>   tflite_model = converter.convert()
> 
> func()
> ```
> 
> Also found some interesting things to bypass this error:
> 
> 1. Change the 1st dim of ipts to something > 1.  or
> 2. use only one filter in conv1d. or
> 3. use another perm=[1,0,2]
> 
> Either of these could eliminate the error. Please stay tuned a bit as I continue to investigate the issue. Thanks.

I tried to follow the converting process with --dump_graphviz_dir and --dump_graphviz_video option in tflite_convert.

I think the error occurred in the following part:
tensorflow/tensorflow/lite/toco/graph_transformations/reorder_reshape_transpose.cc

The tflite_convert computing the 'new_perm' using the function,
const std::vector<int> new_perm =
      ComputeNewPerm(input_dims, intermediate_dims, transpose_op->perm);
, with the following inputs:
input_dims = [1,1,50,2]
intermediate_dims= [1,50,2]
transpose_op->perm = [1,2,50]

I think the function 'ComputeNewPerm' should return 'new_perm' as,
new_perm = [3, 2, 0, 1]
, however when I checked the dot file, it seems it returned
new_perm = [66094776, 3, 0, 1].

66094776 is a garbage value that changes at every run.

I not sure whether the function actually returned incorrect values or not.
Perhaps it returned the correct values but somehow changed afterwards.

But the problem is that the 'reorder_reshape_transpose' seems to created wrong node,
int32[4] = {66094776, 3, 0, 1}
"transpose/perm"
, in the graph transformation.

I hope there exist a solution for this problem.


https://github.com/tensorflow/tensorflow/blob/456fbc0e498e3d10604973de9f46ca48d62267cc/tensorflow/lite/toco/graph_transformations/reorder_reshape_transpose.cc#L217

**Comment by paanguin** [comment]
I made a test code for the 'ComputeNewPerm' function as follows:

`#include <iostream>
#include <unordered_map>
#include <vector>

void show_vector(std::vector<int> vec){
        for (int i; i < vec.size(); i++){
                std::cout << vec[i] << " ";
        }
        std::cout << "\n";
}

template<typename K, typename V>
void print_map(std::unordered_map<K,V> const &m)
{
    for (auto const& pair: m) {
        std::cout << "{" << pair.first << ": " << pair.second << "}, ";
    }
    std::cout << "\n";
}


// Computes a new permutation used to swap a reshape-transpose to a
// transpose-reshape. In this case the permutation operates on the intermediate
// shape.
std::vector<int> ComputeNewPerm(std::vector<int> input_dims,
                                std::vector<int> intermediate_dims,
                                std::vector<int> perm) {
  // These are the major axis of the input.
  std::vector<int> input_indices;
  for (int i = 0; i < input_dims.size(); i++) {
    if (input_dims[i] != 1) {
      input_indices.push_back(i);
    }
  }
  std::cout << "input_indices: ";
  show_vector(input_indices);

  // This maps which indices of the input produced the intermediate indices for
  // non-unary dimensions.
  std::unordered_map<int, int> intermediate_to_input_indices_map;
  for (int i = 0; i < intermediate_dims.size(); i++) {
    if (intermediate_dims[i] != 1) {
      intermediate_to_input_indices_map[i] =
          input_indices[intermediate_to_input_indices_map.size()];  // original

      //intermediate_to_input_indices_map.insert(std::pair<int,int> ( i, input_indices[intermediate_to_input_indices_map.size()] )); // fix


    }
  }
  std::cout << "intermediate_to_input_indices_map: ";
  print_map(intermediate_to_input_indices_map);

  // Translate the transpose permutation to a new permutation starting with the
  // major indices.
  std::vector<int> new_perm;
  new_perm.reserve(input_dims.size());
  for (int i = 0; i < perm.size(); i++) {
    if (intermediate_dims[perm[i]] == 1) continue;

    new_perm.push_back(intermediate_to_input_indices_map[perm[i]]);
  }

  std::cout << "new_perm (0): ";
  show_vector(new_perm);

  // Fill the rest of the transpose in with the ones.
  for (int index = 0; index < input_dims.size(); index++) {
    if (input_dims[index] == 1) {
      new_perm.push_back(index);
    }
  }

  //CHECK_EQ(new_perm.size(), input_dims.size());
  return new_perm;
}

int main() {
  std::vector<int> input_dims{1,1,50,2};
  std::vector<int> intermediate_dims{1,50,2};
  std::vector<int> perm{0,2,1};


  std::cout << "input_dims: ";
  show_vector(input_dims);

  std::cout << "intermediate_dims: ";
  show_vector(intermediate_dims);

  std::cout << "perm: ";
  show_vector(perm);

  const std::vector<int> new_perm =
      ComputeNewPerm(input_dims, intermediate_dims, perm);
  std::cout << "output: " ;
  show_vector(new_perm);
}
`

This gave me following output:

> input_dims: 1 1 50 2
intermediate_dims: 1 50 2
perm: 0 2 1
input_indices: 2 3
intermediate_to_input_indices_map: {2: 0}, {1: 3},
new_perm (0): 0 3
output: 0 3 0 1

**Comment by paanguin** [comment]
> I made a test code for the 'ComputeNewPerm' function to find the problem.
> 
> I think the line with problem was
> 
> ` input_indices[intermediate_to_input_indices_map.size()]; // original`
> 
> https://github.com/tensorflow/tensorflow/blob/456fbc0e498e3d10604973de9f46ca48d62267cc/tensorflow/lite/toco/graph_transformations/reorder_reshape_transpose.cc#L74
> 
> This gave me following output:
> 
> > input_dims: 1 1 50 2
> > intermediate_dims: 1 50 2
> > perm: 0 2 1
> > input_indices: 2 3
> > intermediate_to_input_indices_map: {2: 0}, {1: 3},
> > new_perm (0): 0 3
> > output: 0 3 0 1
> 
> When I change the line with this,
> `intermediate_to_input_indices_map.insert(std::pair<int,int> ( i, input_indices[intermediate_to_input_indices_map.size()] )); // fix`
> , I got following output and I think this is the right version:
> 
> > input_dims: 1 1 50 2
> > intermediate_dims: 1 50 2
> > perm: 0 2 1
> > input_indices: 2 3
> > intermediate_to_input_indices_map: {2: 3}, {1: 2},
> > new_perm (0): 3 2
> > output: 3 2 0 1
> 
> I am trying to change this part of the code, and build / install the tensorflow to see whether this solves the problem or not.

And I confirmed that it solved the problem. Now tflite_convert managed to convert the following graph without error.

`ipts = tf.placeholder(dtype=tf.float32, shape=(None, 50, 50), name='x')
y = tf.layers.conv1d(ipts, filters=2, kernel_size=1)
res = tf.transpose(y, perm=[0,2,1]) 
`

**Comment by paanguin** [comment]
This was my test code.
```

#include <iostream>
#include <unordered_map>
#include <vector>

void show_vector(std::vector<int> vec){
        for (int i; i < vec.size(); i++){
                std::cout << vec[i] << " ";
        }
        std::cout << "\n";
}

template<typename K, typename V>
void print_map(std::unordered_map<K,V> const &m)
{
    for (auto const& pair: m) {
        std::cout << "{" << pair.first << ": " << pair.second << "}, ";
    }
    std::cout << "\n";
}


// Computes a new permutation used to swap a reshape-transpose to a
// transpose-reshape. In this case the permutation operates on the intermediate
// shape.
std::vector<int> ComputeNewPerm(std::vector<int> input_dims,
                                std::vector<int> intermediate_dims,
                                std::vector<int> perm) {
  // These are the major axis of the input.
  std::vector<int> input_indices;
  for (int i = 0; i < input_dims.size(); i++) {
    if (input_dims[i] != 1) {
      input_indices.push_back(i);
    }
  }
  std::cout << "input_indices: ";
  show_vector(input_indices);

  // This maps which indices of the input produced the intermediate indices for
  // non-unary dimensions.
  std::unordered_map<int, int> intermediate_to_input_indices_map;
  for (int i = 0; i < intermediate_dims.size(); i++) {
    if (intermediate_dims[i] != 1) {
      //intermediate_to_input_indices_map[i] =
      //    input_indices[intermediate_to_input_indices_map.size()];  // original

      intermediate_to_input_indices_map.insert(std::pair<int,int> ( i, input_indices[intermediate_to_input_indices_map.size()] )); // fix


    }
  }
  std::cout << "intermediate_to_input_indices_map: ";
  print_map(intermediate_to_input_indices_map);

  // Translate the transpose permutation to a new permutation starting with the
  // major indices.
  std::vector<int> new_perm;
  new_perm.reserve(input_dims.size());
  for (int i = 0; i < perm.size(); i++) {
    if (intermediate_dims[perm[i]] == 1) continue;

    new_perm.push_back(intermediate_to_input_indices_map[perm[i]]);
  }

  std::cout << "new_perm (0): ";
  show_vector(new_perm);

  // Fill the rest of the transpose in with the ones.
  for (int index = 0; index < input_dims.size(); index++) {
    if (input_dims[index] == 1) {
      new_perm.push_back(index);
    }
  }

  //CHECK_EQ(new_perm.size(), input_dims.size());
  return new_perm;
}

int main() {
  std::vector<int> input_dims{1,1,50,2};
  std::vector<int> intermediate_dims{1,50,2};
  std::vector<int> perm{0,2,1};


  std::cout << "input_dims: ";
  show_vector(input_dims);

  std::cout << "intermediate_dims: ";
  show_vector(intermediate_dims);

  std::cout << "perm: ";
  show_vector(perm);

  const std::vector<int> new_perm =
      ComputeNewPerm(input_dims, intermediate_dims, perm);
  std::cout << "output: " ;
  show_vector(new_perm);
}

```

**Comment by KvitnucaZahradka** [comment]
I am using tf version 2.1.0. on Ubuntu 18.04, with the CUDA Version: 10.2 and Driver Version: 440.100. 

Very similar bug appeared. Right after the embedding layer, permutation: perm=[1,0,2] is failing.

The code that fails in 2.1.0 is along the following lines:

```
sze = 77275
negative_samples = tf.keras.Input(shape=(context_size, ), dtype=tf.int64, name='negative_samples')
embedding = tf.keras.layers.Embedding(input_dim=sze,
                                               output_dim=512,
                                               embeddings_initializer='GlorotUniform', name='shared_embedding_layer')

negative_embeddings = embedding(negative_samples)

transposed_embeddings = tf.transpose(negative_embeddings, perm=[1,0,2], name='transpose6')
```

if you attempt to get `transposed_embeddings` out, I am getting an error.

**Comment by KvitnucaZahradka** [comment]
Update: I checked the tf version 2.2.0 and there the transpose seems to work fine.

**Comment by mohantym** [comment]
Hi @lost-libra!
Contrib is removed from latest TF-2.0 version, you need find alternatives for that using this link. Hence you need to modify code according to TF1.X and TF2.X versions.

Could you please try on instructions from [Tflite documentation ](https://www.tensorflow.org/api_docs/python/tf/lite/TFLiteConverter) from  latest stable version of tf 2.5 or 2.4.1 and let us know if this is still an issue.Thanks!
